swarm.go

```diff
@@ -4,187 +4,169 @@ package swarm
+	"errors"
 	"fmt"
-	"io/ioutil"
-	"os"
 	"strings"
 	"sync"
+	"sync/atomic"
 	"time"
 
 	logging "github.com/ipfs/go-log"
 	"github.com/jbenet/goprocess"
 	goprocessctx "github.com/jbenet/goprocess/context"
-	addrutil "github.com/libp2p/go-addr-util"
-	conn "github.com/libp2p/go-libp2p-conn"
-	ci "github.com/libp2p/go-libp2p-crypto"
-	ipnet "github.com/libp2p/go-libp2p-interface-pnet"
 	metrics "github.com/libp2p/go-libp2p-metrics"
-	mconn "github.com/libp2p/go-libp2p-metrics/conn"
 	inet "github.com/libp2p/go-libp2p-net"
 	peer "github.com/libp2p/go-libp2p-peer"
 	pstore "github.com/libp2p/go-libp2p-peerstore"
 	transport "github.com/libp2p/go-libp2p-transport"
 	filter "github.com/libp2p/go-maddr-filter"
-	ps "github.com/libp2p/go-peerstream"
-	pst "github.com/libp2p/go-stream-muxer"
-	tcpt "github.com/libp2p/go-tcp-transport"
-	ws "github.com/libp2p/go-ws-transport"
-	ma "github.com/multiformats/go-multiaddr"
-	psmss "github.com/whyrusleeping/go-smux-multistream"
-	spdy "github.com/whyrusleeping/go-smux-spdystream"
-	yamux "github.com/whyrusleeping/go-smux-yamux"
 	mafilter "github.com/whyrusleeping/multiaddr-filter"
 )
 
-var log = logging.Logger("swarm2")
-
-// PSTransport is the default peerstream transport that will be used by
-// any libp2p swarms.
-var PSTransport pst.Transport
-
-func init() {
-	msstpt := psmss.NewBlankTransport()
-
-	ymxtpt := &yamux.Transport{
-		AcceptBacklog:          8192,
-		ConnectionWriteTimeout: time.Second * 10,
-		KeepAliveInterval:      time.Second * 30,
-		EnableKeepAlive:        true,
-		MaxStreamWindowSize:    uint32(1024 * 512),
-		LogOutput:              ioutil.Discard,
-	}
+// DialTimeout is the maximum duration a Dial is allowed to take.
+// This includes the time between dialing the raw network connection,
+// protocol selection as well the handshake, if applicable.
+var DialTimeout = 60 * time.Second
 
-	msstpt.AddTransport("/yamux/1.0.0", ymxtpt)
-	msstpt.AddTransport("/spdy/3.1.0", spdy.Transport)
+var log = logging.Logger("swarm2")
 
-	// allow overriding of muxer preferences
-	if prefs := os.Getenv("LIBP2P_MUX_PREFS"); prefs != "" {
-		msstpt.OrderPreference = strings.Fields(prefs)
-	}
+// ErrSwarmClosed is returned when one attempts to operate on a closed swarm.
+var ErrSwarmClosed = errors.New("swarm closed")
 
-	PSTransport = msstpt
-}
+// ErrAddrFiltered is returned when trying to register a connection to a
+// filtered address. You shouldn't see this error unless some underlying
+// transport is misbehaving.
+var ErrAddrFiltered = errors.New("address filtered")
 
 // Swarm is a connection muxer, allowing connections to other peers to
 // be opened and closed, while still using the same Chan for all
 // communication. The Chan sends/receives Messages, which note the
 // destination or source Peer.
-//
-// Uses peerstream.Swarm
 type Swarm struct {
-	swarm *ps.Swarm
+	// Close refcount. This allows us to fully wait for the swarm to be torn
+	// down before continuing.
+	refs sync.WaitGroup
```

Comment:
I am not sure if WG is right thing to use for this. If Wait was already called and Add is called after it will panic.

Reply:
Good point, probably easier to read. A waitgroup *should* work as we only call this *after* preventing anything else from starting but yeah, a rwlock is definitely simpler.

Reply:
On further inspection, the code is currently *technically* correct because add only panics if the count reaches zero and then goes above zero). Really, we *want* that behaviour. However, the ref tracking code is still really hard to follow so I'll see what I can do.

Reply:
I tried... there are now more comments. I can't see a nicer (while still efficient) way to do this.